lib/ex_doc/formatter/html.ex

```diff
@@ -147,20 +147,40 @@ defmodule ExDoc.Formatter.HTML do
   end
 
   defp output_setup(build, config) do
-    if File.exists?(build) do
-      build
-      |> File.read!()
-      |> String.split("\n", trim: true)
-      |> Enum.map(&Path.join(config.output, &1))
-      |> Enum.each(&File.rm/1)
-
-      File.rm(build)
-    else
-      File.rm_rf!(config.output)
-      File.mkdir_p!(config.output)
+    safety_path = Path.join(config.output, ".ex_doc")
+
+    cond do
+      File.exists?(build) ->
+        build
+        |> File.read!()
+        |> String.split("\n", trim: true)
+        |> Enum.map(&Path.join(config.output, &1))
+        |> Enum.each(&File.rm/1)
+
+        File.rm(build)
+
+      File.exists?(safety_path) ->
+        File.rm_rf!(config.output)
+        mkdir_output!(config.output)
+
+      not File.exists?(config.output) ->
+        mkdir_output!(config.output)
+
+      true ->
+        raise """
+        ex_doc cannot output to #{config.output}:
+        Directory already exists and is not managed by ex_doc.
+
+        Try giving an unexisting directory as `--output`.
+        """
     end
   end
```

Comment:
Is there a way we could share this functionality? We can either:

1. Move it outside ExDoc.Formatter.HTML and call it earlier (it depends if we can also move the creation of .ex_doc itself higher in the call stack)

2. Make `setup_output` public and invoke it from EPUB (as it does with other HTML functions)

Thoughts?

Reply:
That's a tricky question, I decided to leave them like this intially because I could find enough justifications to separate them, because there are some differences in their logic.

For example, in the EPUB example, we use the original `config.output` value before `normalize_config` appends the project name to it, and the mkdir calls are different in both the HTML and EPUB versions.

Although the error message  could certainly be shared between both, to avoid future inconsistencies.

One approach that might be possible would be making `setup_output` public and then passing the function that does directory and `.ex_doc` file creation as callback, and them on the HTML case, just passing both `path` and `parent` as the same value, what do you think?

Reply:
I think that's worth trying. Do it as additional comments so we revert if necessary?

Reply:
Done, I think it's considerably better now @josevalim